main.cpp

```diff
@@ -38,8 +39,271 @@
+                size_t in2_size, uint8_t *out, size_t max_out_size, unsigned seed)
+{
+    std::srand(seed);
+    if (std::rand() % 2)
+        return insert_part(in1, in1_size, in2, in2_size, out, max_out_size);
+    return overwrite_part(in1, in1_size, in2, in2_size, out, max_out_size);
+}
+
+size_t LLVMFuzzerCustomCrossOver(const uint8_t *in1, size_t in1_size, const uint8_t *in2,
+                               size_t in2_size, uint8_t *out, size_t max_out_size,
+                               unsigned seed)
+{
+    if (in1_size < 9 || in2_size < 9)
+        return 0;
+
+    // Interpret fuzz inputs as strings
+    std::string input1(reinterpret_cast<const char*>(in1), in1_size);
+    std::string input2(reinterpret_cast<const char*>(in2), in2_size);
+
+    // Attempt to bech32 decode the inputs
+    bech32::DecodeResult result1 = bech32::Decode(input1, bech32::CharLimit::CUSTOM_MUTATOR);
+    if (result1.encoding != bech32::Encoding::BECH32) {
+        // Decoding failed
+        return 0;
+    }
+
+    bech32::DecodeResult result2 = bech32::Decode(input2, bech32::CharLimit::CUSTOM_MUTATOR);
+    if (result2.encoding != bech32::Encoding::BECH32) {
+        // Decoding failed
+        return 0;
+    }
+
+    std::string hrp1 = result1.hrp;
+    std::string hrp2 = result2.hrp;
+    std::vector<uint8_t> data1 = result1.data;
+    std::vector<uint8_t> data2 = result2.data;
+
+    std::srand(seed);
+    std::string out_hrp;
+    std::vector<uint8_t> out_data;
+
+    if (std::rand() % 2) {
+        // Cross-over the HRP
+        out_data = data1;
+
+        // Convert strings to vectors for cross_over function
+        std::vector<uint8_t> hrp1_vec(hrp1.begin(), hrp1.end());
+        std::vector<uint8_t> hrp2_vec(hrp2.begin(), hrp2.end());
+        std::vector<uint8_t> out_hrp_vec(max_out_size - data1.size() - 8);
+
+        size_t out_hrp_size = cross_over(
+            hrp1_vec.data(), hrp1_vec.size(),
+            hrp2_vec.data(), hrp2_vec.size(),
+            out_hrp_vec.data(), out_hrp_vec.size(),
+            static_cast<unsigned>(std::rand()));
+
+        // Convert back to string and ensure null termination
+        out_hrp = std::string(out_hrp_vec.begin(), out_hrp_vec.begin() + out_hrp_size);
+    } else {
+        // Cross-over the data part
+        out_hrp = hrp1;
+
+        size_t max_out_data_size = max_out_size - hrp1.size() - 8;
+        std::vector<uint8_t> out_data_vec(max_out_data_size);
+
+        size_t out_data_size = cross_over(
+            data1.data(), data1.size(),
+            data2.data(), data2.size(),
+            out_data_vec.data(), max_out_data_size,
+            (unsigned)std::rand());
+
+        out_data.assign(out_data_vec.begin(), out_data_vec.begin() + out_data_size);
+    }
+
+    // Encode the output
+    std::string encoded = bech32::Encode(bech32::Encoding::BECH32, out_hrp, out_data);
+
+    if (encoded.empty() || encoded.size() > max_out_size) {
```

Comment:
I don't think `encoded` can be empty here.

Reply:
I just removed this check.